### PR TITLE
Continue #1218: Added Mavlink command reception support

### PR DESCRIPTION
**Pull request by TSC21**
This continues the work of @yhabib29 and splits the Mavlink command send and receive functions between two different classes.

### CODE REVIEW

File: src/core/mavlink_commands.cpp [hunk 1]
```diff
@@ -14,20 +14,22 @@ namespace mavsdk {
 //       - The queue used does not support going through and checking each and every
 //         item yet.
 
-MAVLinkCommands::MAVLinkCommands(SystemImpl& parent) : _parent(parent)
+// MavlinkCommandSender class
```
Comment (julianoes) [inline src/core/mavlink_commands.cpp:17]:
```suggestion
```

File: src/core/mavlink_commands.cpp [hunk 2]
```diff
@@ -318,5 +322,101 @@ void MAVLinkCommands::call_callback(
     // we lock ourselves out when we send a command in the callback receiving a command result.
     _parent.call_user_callback([callback, result, progress]() { callback(result, progress); });
 }
+// MavlinkCommandSender class
```
Comment (julianoes) [inline src/core/mavlink_commands.cpp:325]:
```suggestion
```

File: src/core/mavlink_commands.cpp [hunk 3]
```diff
@@ -318,5 +322,101 @@ void MAVLinkCommands::call_callback(
...
+
+// MavlinkCommandReceiver class
```
Comment (julianoes) [inline src/core/mavlink_commands.cpp:327]:
```suggestion
```

I don't like unnecessary comments :smile:

File: src/core/mavlink_commands.cpp [hunk 4]
```diff
@@ -318,5 +322,101 @@ void MAVLinkCommands::call_callback(
...
+    for (auto it = _mavlink_command_handler_table.begin();
+         it != _mavlink_command_handler_table.end();
+         ++it) {
+        if (it->cmd_id == command_int.command) {
+            it->callback(cmd);
+        }
+    }
+}
+
+void MavlinkCommandReceiver::receive_command_long(mavlink_message_t message)
```
Comment (julianoes) [inline src/core/mavlink_commands.cpp:365]: Usually we don't copy the mavlink_message_t's but use `const mavlink_message_t&`. I think that could be done here as well.

File: src/core/mavlink_commands.cpp [hunk 5]
```diff
@@ -318,5 +322,101 @@ void MAVLinkCommands::call_callback(
...
+         it != _mavlink_command_handler_table.end();
+         /* no ++it */) {
+        if (it->cookie == cookie) {
+            it = _mavlink_command_handler_table.erase(it);
+        } else {
+            ++it;
+        }
+    }
+}
+// MavlinkCommandReceiver class
```
Comment (julianoes) [inline src/core/mavlink_commands.cpp:420]:
```suggestion
```

File: src/core/mavlink_commands.h [hunk 6]
```diff
@@ -12,10 +12,97 @@ namespace mavsdk {
 
 class SystemImpl;
 
-class MAVLinkCommands {
+class MavlinkCommandReceiver {
 public:
-    explicit MAVLinkCommands(SystemImpl& parent);
-    ~MAVLinkCommands();
+    explicit MavlinkCommandReceiver(SystemImpl& parent);
```
Comment (julianoes) [inline src/core/mavlink_commands.h:17]:
```suggestion
    explicit MavlinkCommandReceiver(SystemImpl& system_impl);
```

This `parent` is my mistake but we should get away from it. It's wrong.

File: src/core/mavlink_commands.h [hunk 7]
```diff
@@ -12,10 +12,97 @@ namespace mavsdk {
...
+    ~MavlinkCommandReceiver();
+
+    struct Command {
+        uint8_t target_system_id{0};
+        uint8_t target_component_id{0};
+        uint16_t command{0};
+
+        // Int
+        MAV_FRAME frame = MAV_FRAME_GLOBAL_RELATIVE_ALT;
```
Comment (julianoes) [inline src/core/mavlink_commands.h:26]:
```suggestion
        MAV_FRAME frame {MAV_FRAME_GLOBAL_RELATIVE_ALT};
```

as well as the others below.

File: src/core/mavlink_commands.h [hunk 8]
```diff
@@ -12,10 +12,97 @@ namespace mavsdk {
...
+        bool current = 0;
+        bool autocontinue = false;
+
+        // Long
+        uint8_t confirmation = 0;
+
+        // Mixed
+        struct Params {
+            float param1 = NAN;
```
Comment (julianoes) [inline src/core/mavlink_commands.h:35]: Same here.

File: src/core/mavlink_commands.h [hunk 9]
```diff
@@ -12,10 +12,97 @@ namespace mavsdk {
...
+            float param4 = NAN;
+            float param5 = NAN;
+            float param6 = NAN;
+            float param7 = NAN;
+            int32_t x = 0;
+            int32_t y = 0;
+            float z = NAN;
+        } params{};
+
+        Command(mavlink_command_int_t command_int)
```
Comment (julianoes) [inline src/core/mavlink_commands.h:47]: So we now have 3 command representations? `MavlinkCommandSender::CommandLong`, `MavlinkCommandSender::CommandInt` and `MavlinkCommandReceiver::Command`? We should probably find something common, right?

File: tools/rename-enums-to-camelcase.sh [hunk 10]
```diff
@@ -236,14 +236,14 @@ sed -i 's/MAVLinkParameters::Result::CONNECTION_ERROR/MAVLinkParameters::Result:
...
 
-sed -i 's/MAVLinkCommands::Result::SUCCESS/MAVLinkCommands::Result::Success/g' $files
-sed -i 's/MAVLinkCommands::Result::NO_SYSTEM/MAVLinkCommands::Result::NoSystem/g' $files
-sed -i 's/MAVLinkCommands::Result::CONNECTION_ERROR/MAVLinkCommands::Result::ConnectionError/g' $files
-sed -i 's/MAVLinkCommands::Result::BUSY/MAVLinkCommands::Result::Busy/g' $files
-sed -i 's/MAVLinkCommands::Result::COMMAND_DENIED/MAVLinkCommands::Result::CommandDenied/g' $files
-sed -i 's/MAVLinkCommands::Result::TIMEOUT/MAVLinkCommands::Result::Timeout/g' $files
-sed -i 's/MAVLinkCommands::Result::IN_PROGRESS/MAVLinkCommands::Result::InProgress/g' $files
-sed -i 's/MAVLinkCommands::Result::UNKNOWN_ERROR/MAVLinkCommands::Result::UnknownError/g' $files
+sed -i 's/MavlinkCommandSender::Result::SUCCESS/MavlinkCommandSender::Result::Success/g' $files
```
Comment (julianoes) [inline tools/rename-enums-to-camelcase.sh:239]:
```suggestion
sed -i 's/MAVLinkCommands::Result::SUCCESS/MavlinkCommandSender::Result::Success/g' $files
```

Same below.

File: src/core/system_impl.h [hunk 11]
```diff
@@ -236,6 +236,13 @@ class SystemImpl : public Sender {
     SystemImpl(const SystemImpl&) = delete;
     const SystemImpl& operator=(const SystemImpl&) = delete;
 
+    void register_mavlink_command_handler(
+        uint16_t cmd_id,
+        MavlinkCommandReceiver::mavlink_command_handler_t callback,
+        const void* cookie);
+    void unregister_mavlink_command_handler(uint16_t cmd_id, const void* cookie);
+    void unregister_all_mavlink_command_handlers(const void* cookie);
```
Comment (julianoes) [inline src/core/system_impl.h:244]: As I understand it, we are not using this interface just yet but in the future there will be a new plugin making use of it, and that one will be tested. Right?

**Review by julianoes — CHANGES_REQUESTED** [review]

File: src/core/mavlink_commands.h (hunk 9 above)
Comment (TSC21) [reply src/core/mavlink_commands.h:47]: Yeah probably. Any suggestion? Maybe two parent classes, `CommandLong` and `CommandInt`?

File: tools/rename-enums-to-camelcase.sh (hunk 10 above)
Comment (TSC21) [reply tools/rename-enums-to-camelcase.sh:239]: @julianoes but didn't we rename `MAVLinkCommands` to `MavlinkCommandSender`? Or do we want a `MAVLinkCommands` class?

File: src/core/system_impl.h (hunk 11 above)
Comment (TSC21) [reply src/core/system_impl.h:244]: Yes that is correct

File: tools/rename-enums-to-camelcase.sh (hunk 10 above)
Comment (julianoes) [reply tools/rename-enums-to-camelcase.sh:239]: This script is to rename from version < v0.25.0 to v0.25.0. At that point `MAVLinkCommands` was still called `MAVLinkCommands`.

File: src/core/mavlink_commands.h (hunk 9 above)
Comment (TSC21) [reply src/core/mavlink_commands.h]: I have made a clear separation on the `MavlinkCommandReceiver`. We now have 4 command representations (structs), but the way they are handle are a bit different, i.e. `MavlinkCommandSender::CommandLong` is different from `MavlinkCommandReceiver::CommandLong` (one is created to send the command, the other parses a command and stores the data).

File: tools/rename-enums-to-camelcase.sh (hunk 10 above)
Comment (TSC21) [reply tools/rename-enums-to-camelcase.sh:239]: Ah sorry ok

**Review by julianoes — APPROVED** [review]

**Review by JonasVautherin — APPROVED** [review]

**Pull request merged**